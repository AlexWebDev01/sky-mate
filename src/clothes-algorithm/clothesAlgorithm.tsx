import { TEMPERATURE_RANGES, CLOTHES_ADVICES } from "./clothesAlgorithmData";

const getTempRange = (temp: number): string => {
  for (let i = 0; i < TEMPERATURE_RANGES.length; i++) {
    if (temp < TEMPERATURE_RANGES[i].highestTemp) {
      return TEMPERATURE_RANGES[i].name;
    }
  }

  return "";
};

export const calculateClothesAdvice = (
  weatherCondition: string,
  temp: number
) => {
  const clothesAdvices = CLOTHES_ADVICES[weatherCondition];
  const tempRangeName = getTempRange(temp);

  if (clothesAdvices[tempRangeName]) {
    return { adviceData: clothesAdvices[tempRangeName], tempRangeName };
  } else {
    console.log(`For now there is no such conditions. 
        Temperature: ${temp}, 
        Temperature range: ${tempRangeName}, 
        Weather condition: ${weatherCondition}
        `);

    return null;
  }
};
