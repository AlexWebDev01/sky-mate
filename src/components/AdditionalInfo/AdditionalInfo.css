@import url('@shared/assets/fonts/fonts.css');
.additional-info {
  text-align: left;
  margin-bottom: 40px;
}
.additional-info .title {
  font-family: Circe-medium, sans-serif;
  font-size: 16px;
}
.additional-info .rain,
.additional-info .snow {
  color: var(--blue);
}
.additional-info .sun,
.additional-info .clear {
  color: var(--orange);
}
.additional-info .clouds,
.additional-info .fog,
.additional-info .dust,
.additional-info .mist {
  color: var(--green);
}
.additional-info .rain-light,
.additional-info .snow-light {
  color: var(--light-blue);
}
.additional-info .sun-light,
.additional-info .clear-light {
  color: var(--light-orange);
}
.additional-info .clouds-light,
.additional-info .fog-light,
.additional-info .dust-light,
.additional-info .mist-light {
  color: var(--light-green);
}
.additional-info .value {
  font-family: Oddval;
  font-size: 16px;
  font-weight: 600;
} /*# sourceMappingURL=AdditionalInfo.css.map */
