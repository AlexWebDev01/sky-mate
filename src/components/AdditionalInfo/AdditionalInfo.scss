@import url('../../shared/assets/fonts/fonts.css');

.additional-info {
  text-align: left;
  margin-bottom: 40px;

  .title {
    font-family: Circe-medium, sans-serif;
    font-size: 16px;
  }

  .rain,
  .snow {
    color: var(--blue);
  }

  .sun,
  .clear {
    color: var(--orange);
  }

  .clouds,
  .fog,
  .dust,
  .mist {
    color: var(--green);
  }

  .rain-light,
  .snow-light {
    color: var(--light-blue);
  }

  .sun-light,
  .clear-light {
    color: var(--light-orange);
  }

  .clouds-light,
  .fog-light,
  .dust-light,
  .mist-light {
    color: var(--light-green);
  }

  .value {
    font-family: Oddval;
    font-size: 16px;
    font-weight: 600;
  }
}
