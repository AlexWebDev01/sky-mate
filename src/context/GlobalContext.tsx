import { useState, createContext, useContext, FunctionComponent } from "react";
import {
  Coordinates,
  GlobalContext,
  GlobalProviderProps,
} from "./GlobalContext.interface";

const GlobalContext = createContext<GlobalContext | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error("useWords must be used within a WordsProvider");
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);
  const [weatherData, setWeatherData] = useState<object | null>(null);
  const [location, setLocation] = useState<string | null>(null);
  const [pageStyle, setPageStyle] = useState<string>("");

  const sharedState: GlobalContext = {
    coordinates,
    setCoordinates,
    weatherData,
    setWeatherData,
    location,
    setLocation,
    pageStyle,
    setPageStyle,
  };

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
