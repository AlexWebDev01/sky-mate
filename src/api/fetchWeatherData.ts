import { Coordinates } from "../context/GlobalContext.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: string
) => {
  console.log("COORDINATES:", coordinates);

  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data = await response.json();

  return data;
};
