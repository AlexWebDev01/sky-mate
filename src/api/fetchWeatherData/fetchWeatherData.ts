import { Units, Coordinates, WeatherData } from "./fetchWeatherData.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: Units
): Promise<WeatherData> => {
  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data: WeatherData = await response.json();

  return data;
};
