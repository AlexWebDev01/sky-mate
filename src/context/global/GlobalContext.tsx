import {
  useState,
  createContext,
  useContext,
  FunctionComponent,
  useCallback,
  useMemo,
} from 'react';
import {
  GlobalContext as GlobalContextInterface,
  GlobalProviderProps,
  GlobalState,
} from './GlobalContext.interface';

import { fetchLocationData } from '@api/fetchLocationData/fetchLocationData';
import { fetchUserLocatioByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP';
import { separateCoordinates } from '@shared/helpers/date/date';
import { fetchWeatherData } from '@api/fetchWeatherData/fetchWeatherData';
import { calculateClothesAdvice } from '@shared/helpers/clothesAlgorithm';

import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import { LocationDataByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP.interface';
import { LocationData } from '@api/fetchLocationData/fetchLocationData.interface';
import {
  Units,
  Coordinates,
} from '@api/fetchWeatherData/fetchWeatherData.interface';

const GlobalContext = createContext<GlobalContextInterface | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Global context should be used in GlobalProvider!');
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [state, setState] = useState<GlobalState>({
    isLoading: true,
    coordinates: null,
    location: null,
    weatherData: null,
    pageStyle: '',
    clothesAdvice: null,
  });
  const [error, setError] = useState<string>('');

  const getWeatherData = useCallback(async (coordinates: Coordinates) => {
    const data = await fetchWeatherData(coordinates, Units.metric);
    const weatherCondition =
      data.daily[0].weather[0].main.toLowerCase() as WeatherConditions;
    const averageTemp = data.daily[0].temp.day;
    const clothesAdvice = calculateClothesAdvice(weatherCondition, averageTemp);
    return {
      weatherData: data,
      clothesAdvice,
      pageStyle: weatherCondition,
    };
  }, []);

  const fetchData = useCallback(async () => {
    console.group('FETCH DATA');
    console.info('Started initial data fetching');

    setState((prevState) => ({ ...prevState, isLoading: true }));

    try {
      const locationData: LocationDataByIP = await fetchUserLocatioByIP();
      const coordinates = separateCoordinates(locationData.loc);

      const { weatherData, clothesAdvice, pageStyle } =
        await getWeatherData(coordinates);
      setState((prevState) => ({
        ...prevState,
        weatherData,
        clothesAdvice,
        pageStyle,
        coordinates,
        location: locationData.city,
        isLoading: false,
      }));

      console.info('Fetched & calculated data: ', {
        weatherData,
        clothesAdvice,
        pageStyle,
        city: locationData.city,
      });
      console.info('Finished initial data fetching');
      console.groupEnd();
    } catch (error) {
      if (error instanceof Error) {
        setError('Oops, unexpected error!');
      }

      console.error('Failed to fetch initial data:', error);
    }
  }, [getWeatherData]);

  const handleSearch = useCallback(
    async (location: string) => {
      try {
        console.group('HANDLE SEARCH');
        console.info('Started fetching data for location:', location);
        const locationData: LocationData = await fetchLocationData(location);
        const { name, lat, lon } = locationData;

        const { weatherData, clothesAdvice, pageStyle } = await getWeatherData({
          lat,
          lon,
        });

        setState((prevState) => ({
          ...prevState,
          coordinates: { lat, lon },
          location: name,
          weatherData,
          clothesAdvice,
          pageStyle,
        }));
        console.info('Fetched & calculated data: ', {
          weatherData,
          clothesAdvice,
          pageStyle,
          city: name,
        });
        console.info('Finished fetching data for location:', location);
      } catch (error) {
        if (error instanceof Error && error.message === 'Location not found') {
          setError('Location not found');
        } else if (error instanceof Error) {
          setError('Oops, unexpected error!');
        }

        console.error('Error fetching data for location:', error);
      }
    },
    [getWeatherData]
  );

  const sharedState = useMemo<GlobalContextInterface>(
    () => ({
      state,
      error,
      setError,
      fetchData,
      handleSearch,
    }),
    [state, error, setError, fetchData, handleSearch]
  );

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
