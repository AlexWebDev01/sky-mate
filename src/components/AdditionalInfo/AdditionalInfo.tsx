import { useGlobalContext } from '@context/global/GlobalContext';
import { epochToLocalTime } from '@shared/helpers/date/date';

import './AdditionalInfo.scss';

interface Props {
  isMainPage: boolean;
  expanded: boolean;
}

const AdditionalInfo = ({ isMainPage, expanded }: Props) => {
  const { state } = useGlobalContext();
  const { weatherData } = state;

  if (!weatherData) {
    return null;
  }

  const additionalInfoContent = [
    {
      title: 'sunrise',
      value: epochToLocalTime(
        weatherData.current.sunrise,
        weatherData.timezone_offset
      ),
    },
    {
      title: 'sunset',
      value: epochToLocalTime(
        weatherData.current.sunset,
        weatherData.timezone_offset
      ),
    },
    {
      title: 'wind speed',
      value: `${Math.round(weatherData.current.wind_speed * 10) / 10} m/s`,
    },
    { title: 'pressure', value: `${weatherData.current.pressure} m/s` },
  ];

  const weatherCondition = weatherData.daily[0].weather[0].main.toLowerCase();
  const titleStyle = isMainPage ? weatherCondition + '-light' : '';
  const contentStyle = isMainPage ? '' : weatherCondition;

  return (
    <div className={expanded ? 'additional-info' : 'hide'}>
      {additionalInfoContent.map((item, index) => {
        return (
          <div key={index}>
            <div className={`${titleStyle} title`}>{item.title}</div>
            <div className={`${contentStyle} value`}>{item.value}</div>
          </div>
        );
      })}
    </div>
  );
};

export default AdditionalInfo;
