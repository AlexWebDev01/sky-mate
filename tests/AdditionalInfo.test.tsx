import '@testing-library/jest-dom';
import { queryByTestId, render, screen } from '@testing-library/react';
import { GlobalProvider } from '@context/global/GlobalContext';
import AdditionalInfo from '@components/AdditionalInfo/AdditionalInfo';
import { WeatherData } from '@api/fetchWeatherData/fetchWeatherData.interface';
import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';

const mockWeatherData: WeatherData = {
  current: {
    clouds: 75,
    dew_point: 10,
    dt: 1631990400,
    feels_like: 12,
    humidity: 50,
    pressure: 1000,
    sunrise: 1631990400,
    sunset: 1632034400,
    temp: 15,
    uvi: 5,
    visibility: 10000,
    weather: [
      {
        main: WeatherConditions.clouds,
        description: 'overcast clouds',
        icon: '04d',
        id: 804,
      },
    ],
    wind_deg: 180,
    wind_speed: 2.5,
  },
  daily: [
    {
      dt: 1631990400,
      sunrise: 1631990400,
      sunset: 1632034400,
      moonrise: 1631990400,
      moonset: 1632034400,
      moon_phase: 0.5,
      temp: { day: 15, min: 10, max: 20, night: 10, eve: 12, morn: 10 },
      feels_like: { day: 12, night: 10, eve: 12, morn: 10 },
      pressure: 1000,
      humidity: 50,
      dew_point: 10,
      wind_speed: 2.5,
      wind_deg: 180,
      wind_gust: 3.5,
      clouds: 75,
      uvi: 5,
      pop: 0.5,
      rain: 0.5,
      summary: 'overcast clouds',
      weather: [
        {
          main: WeatherConditions.sun,
          description: 'clear sky',
          icon: '01d',
          id: 800,
        },
      ],
    },
  ],
  hourly: [
    {
      dt: 1631990400,
      temp: 15,
      feels_like: 12,
      pressure: 1000,
      humidity: 50,
      dew_point: 10,
      uvi: 5,
      clouds: 75,
      visibility: 10000,
      wind_speed: 2.5,
      wind_deg: 180,
      wind_gust: 3.5,
      pop: 0.5,
      weather: [
        {
          main: WeatherConditions.clouds,
          description: 'overcast clouds',
          icon: '04d',
          id: 804,
        },
      ],
    },
  ],
  lat: 56.946,
  lon: 24.1059,
  timezone: 'Europe/Riga',
  timezone_offset: 10800,
};

jest.mock('@context/global/GlobalContext', () => ({
  useGlobalContext: jest.fn(),
  GlobalProvider: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));

describe('AdditionalInfo', () => {
  beforeEach(() => {
    const { useGlobalContext } = jest.requireMock(
      '@context/global/GlobalContext'
    );
    useGlobalContext.mockImplementation(() => ({
      state: {
        weatherData: mockWeatherData,
        isLoading: false,
        coordinates: { lat: 56.946, lon: 24.1059 },
        location: 'Riga',
        pageStyle: 'clouds',
        clothesAdvice: 'Wear a jacket',
      },
      error: '',
      setError: jest.fn(),
      fetchData: jest.fn(),
      handleSearch: jest.fn(),
    }));
  });

  test('renders without crashing', () => {
    render(
      <GlobalProvider>
        <AdditionalInfo isMainPage={true} expanded={false} />
      </GlobalProvider>
    );
    expect(screen.getByTestId('additional-info')).toBeInTheDocument();
  });

  test('does not render if weatherData is null', () => {
    const { useGlobalContext } = jest.requireMock(
      '@context/global/GlobalContext'
    );
    useGlobalContext.mockImplementation(() => ({
      state: {
        weatherData: null,
        isLoading: false,
        coordinates: { lat: 56.946, lon: 24.1059 },
        location: 'Riga',
        pageStyle: 'clouds',
        clothesAdvice: 'Wear a jacket',
      },
      error: '',
      setError: jest.fn(),
      fetchData: jest.fn(),
      handleSearch: jest.fn(),
    }));

    const { container } = render(
      <GlobalProvider>
        <AdditionalInfo isMainPage={true} expanded={false} />
      </GlobalProvider>
    );
    expect(queryByTestId(container, 'additional-info')).toBeNull();
  });
});
