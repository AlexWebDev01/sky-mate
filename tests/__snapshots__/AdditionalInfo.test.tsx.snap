// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInfo matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="hide"
    data-testid="additional-info"
  >
    <div>
      <div
        class="sun-light title"
      >
        sunrise
      </div>
      <div
        class=" value"
      >
        21:40
      </div>
    </div>
    <div>
      <div
        class="sun-light title"
      >
        sunset
      </div>
      <div
        class=" value"
      >
        09:53
      </div>
    </div>
    <div>
      <div
        class="sun-light title"
      >
        wind speed
      </div>
      <div
        class=" value"
      >
        2.5 m/s
      </div>
    </div>
    <div>
      <div
        class="sun-light title"
      >
        pressure
      </div>
      <div
        class=" value"
      >
        1000 m/s
      </div>
    </div>
  </div>
</DocumentFragment>
`;
